apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/jdbc/StatementInstrumentation.java

```diff
@@ -0,0 +1,94 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.jdbc;
+
+import co.elastic.apm.bci.ElasticApmInstrumentation;
+import co.elastic.apm.bci.VisibleForAdvice;
+import co.elastic.apm.impl.ElasticApmTracer;
+import co.elastic.apm.impl.transaction.Span;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import javax.annotation.Nullable;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+import static net.bytebuddy.matcher.ElementMatchers.isInterface;
+import static net.bytebuddy.matcher.ElementMatchers.isPublic;
+import static net.bytebuddy.matcher.ElementMatchers.isSubTypeOf;
+import static net.bytebuddy.matcher.ElementMatchers.nameContains;
+import static net.bytebuddy.matcher.ElementMatchers.nameStartsWith;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+
+/**
+ * Creates spans for JDBC calls
+ */
+public class StatementInstrumentation extends ElasticApmInstrumentation {
+
+    @Nullable
+    private static ElasticApmTracer tracer;
+    @Nullable
+    private static ApmJdbcEventListener jdbcEventListener;
+
+    // not inlining as we can then set breakpoints into this method
+    // also, we don't have class loader issues when doing so
+    // another benefit of not inlining is that the advice methods are included in coverage reports
+    @Nullable
+    @VisibleForAdvice
+    @Advice.OnMethodEnter(inline = false)
+    public static Span onBeforeExecute(@Advice.This Statement statement, @Advice.Argument(0) String sql) throws SQLException {
+        if (tracer != null && jdbcEventListener != null) {
+            return jdbcEventListener.createJdbcSpan(sql, statement.getConnection(), tracer.currentSpan());
+        }
+        return null;
+    }
+
+
+    @VisibleForAdvice
+    @Advice.OnMethodExit(inline = false, onThrowable = SQLException.class)
+    public static void onAfterExecute(@Advice.Enter @Nullable Span span, @Advice.Thrown SQLException e) {
+        if (span != null) {
+            span.end();
+        }
+    }
+
+    @Override
+    public void init(ElasticApmTracer tracer) {
+        StatementInstrumentation.tracer = tracer;
+        StatementInstrumentation.jdbcEventListener = new ApmJdbcEventListener(tracer);
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return not(isInterface())
+            .and(nameContains("Statement"))
```

Comment:
Just out of curiosity, why was this needed?  Is this a performance optimization to avoid the `isSubTypeOf` check?

Reply:
See the comment in https://github.com/elastic/apm-agent-java/blob/master/apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/servlet/ServletInstrumentation.java#L69
Basically it’s because the isSubType marcher can be pretty expensive. It causes a lot of type descriptions to be parsed by ByteBuddy. I stagemonitor, I cached TypeDescriptions to improve instrumentation performance (startup times). A user experienced an OOME because of an isSubType matcher. But I should re evaluate with a recent Byte Buddy version.